### ISSUE TITLE
how to start the program

### ISSUE THREAD

**Issue by lxyzler** [leondgarse/Keras_insightface#11]
How can I test with pictures or videos to see the effect

**Comment by leondgarse** [comment]
I just uploaded a new test file `video_test.py`
- Usage
```sh
$ CUDA_VISIBLE_DEVICES='-1' python video_test.py -h
usage: video_test.py [-h] -m MODEL_FILE [-k KNOWN_USER] [-K KNOWN_USER_FORCE]
                     [-b EMBEDDING_BATCH_SIZE] [-s VIDEO_SOURCE]
                     [-t DIST_THRESH] [-p FRAMES_PER_DETECT]

optional arguments:
  -h, --help            show this help message and exit
  -m MODEL_FILE, --model_file MODEL_FILE
                        Saved basic_model file path, NOT model (default: None)
  -k KNOWN_USER, --known_user KNOWN_USER
                        Folder containing user images data (default: None)
  -K KNOWN_USER_FORCE, --known_user_force KNOWN_USER_FORCE
                        Folder containing user images data, force reload
                        (default: None)
  -b EMBEDDING_BATCH_SIZE, --embedding_batch_size EMBEDDING_BATCH_SIZE
                        Batch size for extracting known user embedding data
                        (default: 4)
  -s VIDEO_SOURCE, --video_source VIDEO_SOURCE
                        Video source (default: 0)
  -t DIST_THRESH, --dist_thresh DIST_THRESH
                        Cosine dist thresh, dist lower than this will be
                        Unknown (default: 0.6)
  -p FRAMES_PER_DETECT, --frames_per_detect FRAMES_PER_DETECT
                        Do detect every [NUM] frame (default: 5)
```
```sh
CUDA_VISIBLE_DEVICES='-1' ./video_test.py -m './checkpoints/mobilenet_adamw_BS256_E80_arc_tripD_basic_agedb_30_epoch_123_0.955333.h5' -k temp_test/
```
- `MODEL_FILE` is the trained keras basic_model.
- `KNOWN_USER` indicates a folder including images just in the same format as `train dataset`.
  ```sh
  .               # dataset folder
  ├── user_1       # person folder
  │   ├── 100.jpg # face image
  │   ├── 101.jpg # face image
  │   └── 102.jpg # face image
  ├── user_2       # person folder
  │   ├── 111.jpg
  │   ├── 112.jpg
  │   └── 113.jpg
  ├── user_3
  │   ├── 707.jpg
  │   ├── 708.jpg
  │   └── 709.jpg
  ```
- `KNOWN_USER_FORCE` used for reload from `KNOWN_USER`, or it will use a cached `npz` file. Use this if changed model or user data.
- It also has a function `image_recognize`, which is used to detect and recognize faces in a single image, you can read it for the usage.

**Comment by cuongdv1** [comment]
@leondgarse 
I am trying to reproduce your env
Could you share settings to get 99.xx% accuracy of mobilenet?

**Comment by leondgarse** [comment]
@cuongdv1 As I remember, it's something like this:
```py
import losses, train

data_path = '/datasets/faces_emore_112x112_folders'
eval_paths = ['/datasets/faces_emore/lfw.bin', '/datasets/faces_emore/cfp_fp.bin', '/datasets/faces_emore/agedb_30.bin']

basic_model = train.buildin_models("mobilenet", dropout=0, emb_shape=256, output_layer='GDC')
tt = train.Train(data_path, save_path='keras_mobilenet_emore_nadam_soft_arc_tripD.h5', eval_paths=eval_paths,
    basic_model=basic_model, lr_base=0.001, batch_size=512, random_status=2)
sch = [
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "epoch": 15, "optimizer": "nadam"},
  {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 30},
  {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.35},
  {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.3},
  {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.25},
  {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.2},
]
tt.train(sch, 0)
```
You can also refer to this issue [MobileFacenet SE Train from scratch #9](https://github.com/leondgarse/Keras_insightface/issues/9). I can also test it again later.

**Comment by cuongdv1** [comment]
@leondgarse 
Thank you very much for sharing,
I am reproducing it,
But about ArcFaceLossT4, what is difference? comparing to ArcFaceLoss, 
I read paper and MXnet code but not found information 
could you explain it?

Thank you in advance

**Comment by leondgarse** [comment]
It's used to be here [insightface/src/train.py](https://github.com/deepinsight/insightface/blob/18d67be1eb376738964c2cabfa87590ae496c15e/src/train.py#L330), and it's the recommended one. but in recently update, it's gone...

It's just a simplified version of `ArcfaceLoss` where `margin1=1.0, margin3=0.0` and `margin2=margin`. Though looks different, but should have the same effect.

**Comment by cuongdv1** [comment]
> @cuongdv1 As I remember, it's something like this:
> 
> ```python
> import losses, train
> 
> data_path = '/datasets/faces_emore_112x112_folders'
> eval_paths = ['/datasets/faces_emore/lfw.bin', '/datasets/faces_emore/cfp_fp.bin', '/datasets/faces_emore/agedb_30.bin']
> 
> basic_model = train.buildin_models("mobilenet", dropout=0, emb_shape=256, output_layer='GDC')
> tt = train.Train(data_path, save_path='keras_mobilenet_emore_nadam_soft_arc_tripD.h5', eval_paths=eval_paths,
>     basic_model=basic_model, lr_base=0.001, batch_size=512, random_status=2)
> sch = [
>   {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "epoch": 15, "optimizer": "nadam"},
>   {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 30},
>   {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.35},
>   {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.3},
>   {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.25},
>   {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.2},
> ]
> tt.train(sch, 0)
> ```
> 
> You can also refer to this issue [MobileFacenet SE Train from scratch #9](https://github.com/leondgarse/Keras_insightface/issues/9). I can also test it again later.
> Also, you can raise a new issue, this is totally a different one...

@leondgarse 
I reproduced above training schedule, but accuracy results are not same as yours
my accuracy: LFW: 99.51%, **CFP_FP: 90.98%**
yours: LFW: 99.61, **CFP_FP: 94.81**

CFP-FP accuracy is quite low compared to yours,
Do you have any idea? or any parameter setting or training schedule should be tried?

**Comment by leondgarse** [comment]
@cuongdv1 
- Ya, you result looks not a good one... I have also tested just the same script above using `batch_size=480`, it's plotted the orange line. The best result is `Epoch 72: lfw - 0.994667, cfp_fp - 0.933714, agedb - 0.950667`.
- The blue line is the result I published in the `readme`. As you can see, It's trained `80 epochs softmax + centerloss` --> `2 epochs bottleneckonly` --> `50 epochs arcface + triplet`, and `batch_size=256`。
![Figure_4](https://user-images.githubusercontent.com/5744524/100830742-4b239680-349f-11eb-8287-113080d9f0a7.png)

This is the script, the reason I'm not sharing it at first place is that:
- I think this process is too complicated...
- I changed `CenterLoss` since then, replaced `reduce_mean` with `reduce_sum` on [losses.py#L199](https://github.com/leondgarse/Keras_insightface/blob/b8e17ae9960e8069e4655c55c4b30c8a34e06773/losses.py#L199), which is learned from others implementation. This enlarged `CenterLoss`, so the `"centerloss"` parameter should be changed.
```py
import losses, train
import tensorflow_addons as tfa

data_path = '/datasets/faces_emore_112x112_folders'
eval_paths = ['/datasets/faces_emore/lfw.bin', '/datasets/faces_emore/cfp_fp.bin', '/datasets/faces_emore/agedb_30.bin']

basic_model = train.buildin_models("mobilenet", dropout=0, emb_shape=256, output_layer='GDC')
tt = train.Train(data_path, save_path='keras_mobilenet_emore_adamw_5e5_soft_centerD_arc_tripD.h5', eval_paths=eval_paths,
    basic_model=basic_model, lr_base=0.001, batch_size=256, random_status=3)
optimizer = tfa.optimizers.AdamW(learing_rate=0.001, weight_decay=5e-5)
sch = [
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 1, "epoch": 10, "optimizer": optimizer},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 10, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 20, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 30, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 40, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 50, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 60, "epoch": 10},
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "centerloss": 70, "epoch": 10},
  {"loss": losses.ArcfaceLoss(scale=64), "bottleneckOnly": True, "epoch": 2},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.35},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.3},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.25},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 64, "alpha": 0.2},
]
tt.train(sch, 0)
```
